src/core/pointcloud/qgspointcloudlayerprofilegenerator.cpp

```diff
@@ -0,0 +1,554 @@
+  }
+  return res;
+}
+
+QVector<QgsGeometry> QgsPointCloudLayerProfileResults::asGeometries() const
+{
+  // TODO -- cache?
+  QVector< QgsGeometry > res;
+  res.reserve( results.size() );
+  for ( const PointResult &point : results )
+  {
+    res.append( QgsGeometry( new QgsPoint( point.x, point.y, point.z ) ) );
+  }
+  return res;
+}
+
+QgsDoubleRange QgsPointCloudLayerProfileResults::zRange() const
+{
+  return QgsDoubleRange( minZ, maxZ );
+}
+
+void QgsPointCloudLayerProfileResults::renderResults( QgsProfileRenderContext &context )
+{
+  QPainter *painter = context.renderContext().painter();
+  if ( !painter )
+    return;
+
+  const QgsScopedQPainterState painterState( painter );
+
+  painter->setBrush( Qt::NoBrush );
+  painter->setPen( Qt::NoPen );
+
+  switch ( pointSymbol )
+  {
+    case Qgis::PointCloudSymbol::Square:
+      // for square point we always disable antialiasing -- it's not critical here and we benefit from the performance boost disabling it gives
+      context.renderContext().painter()->setRenderHint( QPainter::Antialiasing, false );
+      break;
+
+    case Qgis::PointCloudSymbol::Circle:
+      break;
+  }
+
+  const double minDistance = context.distanceRange().lower();
+  const double maxDistance = context.distanceRange().upper();
+  const double minZ = context.elevationRange().lower();
+  const double maxZ = context.elevationRange().upper();
+
+  const QRectF visibleRegion( minDistance, minZ, maxDistance - minDistance, maxZ - minZ );
+  QPainterPath clipPath;
+  clipPath.addPolygon( context.worldTransform().map( visibleRegion ) );
+  painter->setClipPath( clipPath, Qt::ClipOperation::IntersectClip );
+
+  const double penWidth = context.renderContext().convertToPainterUnits( pointSize, pointSizeUnit );
+
+  for ( const PointResult &point : std::as_const( results ) )
+  {
+    QPointF p = context.worldTransform().map( QPointF( point.distance, point.z ) );
+    QColor color = pointColor;
+    if ( opacityByDistanceEffect )
+      color.setAlphaF( color.alphaF() * ( 1.0 - std::pow( point.curveDistance / tolerance, 0.5 ) ) );
+
+    switch ( pointSymbol )
+    {
+      case Qgis::PointCloudSymbol::Square:
+        painter->fillRect( QRectF( p.x() - penWidth * 0.5,
+                                   p.y() - penWidth * 0.5,
+                                   penWidth, penWidth ), color );
+        break;
+
+      case Qgis::PointCloudSymbol::Circle:
+        painter->setBrush( QBrush( color ) );
+        painter->setPen( Qt::NoPen );
+        painter->drawEllipse( QRectF( p.x() - penWidth * 0.5,
+                                      p.y() - penWidth * 0.5,
+                                      penWidth, penWidth ) );
+        break;
+    }
+  }
+}
+
+QgsProfileSnapResult QgsPointCloudLayerProfileResults::snapPoint( const QgsProfilePoint &point, const QgsProfileSnapContext &context )
+{
+  QgsProfileSnapResult result;
+  Q_UNUSED( point )
+  Q_UNUSED( context )
+#if 0
+  // TODO -- index
```

Comment:
Is it too slow without index to be usable? but yeah, building an index on first snapPoint() would make snapping really fast :-)

Reply:
I've implemented the index & snapping now :smile: